src/Symfony/Component/Form/DependencyInjection/FormPass.php

```diff
@@ -78,18 +85,20 @@ public function process(ContainerBuilder $container)
             }
 
             $typeExtensions[$extendedType][] = $serviceId;
+
+            // Add form type extension service to the service locator
+            $servicesMap[$serviceId] = new Reference($serviceId);
         }
 
         $definition->replaceArgument(2, $typeExtensions);
 
         $guessers = array_keys($container->findTaggedServiceIds($this->formTypeGuesserTag));
         foreach ($guessers as $serviceId) {
-            $serviceDefinition = $container->getDefinition($serviceId);
-            if (!$serviceDefinition->isPublic()) {
-                throw new InvalidArgumentException(sprintf('The service "%s" must be public as form type guessers are lazy-loaded.', $serviceId));
-            }
+            // Add form type guesser service to the service locator
+            $servicesMap[$serviceId] = new Reference($serviceId);
         }
 
+        $definition->replaceArgument(0, new ServiceLocatorArgument($servicesMap));
```

Comment:
I'm wondering if mixing types/guessers/extensions in the same locator is a good thing here, as we have still 3 constructor arguments that's a bit confusing to me.
Instead, couldn't `guesserServices` be an `IteratorArgument` of guessers? It seems we only iterate over them to build a chain one, no need for having them identified right?